---
title: "Annotate S288C UTRs from Pelechano 2013 data"
author: "Edward Wallace, Edward.Wallace@ed.ac.uk"
date: "24 March 2020"
output: 
    html_document:
        toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(here)
```

# Functions to read, write, manipulate gffs

```{r read_functions}
read_gff <- function(file){
    # tidyverse read gff function from rmonad vignette
    # https://cran.r-project.org/web/packages/rmonad/vignettes/gff-processing.html
    
    readr::read_tsv(
        file,
        col_names = c(
            "Seqid",
            "Source",
            "Type",
            "Start",
            "End",
            "Score",
            "Strand",
            "Phase",
            "Attr"
        ),
        na        = ".",
        comment   = "#",
        col_types = "ccciidcic"
    )
}

write_gff <- function(x,path,...) {
    x %>%
        dplyr::ungroup() %>%
        dplyr::select(Seqid,Source,Type,Start,End,Score,Strand,Phase,Attr) %>%
        tidyr::replace_na(list(Score=".",Strand=".",Phase=".")) %>%
        readr::write_tsv(path=path,...)
}

gene_from_attr <- function(attr_string,
                           before="Name=",gene_str="[\\w-]+",after="_") {
    # pull a gene name from an attribute string
    match_str = paste0(before,gene_str,after)
    attr_string %>%
        str_extract(match_str) %>%
        str_remove(before) %>%
        str_remove(after)
}
# test_bit <- gff_CDS$attr[1]
# gene_from_attr(test_bit)

assign_5UTR <- function(gff_join_row,defaultlength=25) {
    Attr.tx <- pull(gff_join_row,Attr.tx)
    if( !is.na(Attr.tx) ) {
        Attr.UTR <- Attr.tx %>%
            str_extract("ID=[\\w-_]+") %>%
            str_remove("ID=") %>% 
            paste0("ID=",.,"_5UTR;Parent=",.) 
    } else {
        Attr.UTR <- pull(gff_join_row,Gene) %>%
            paste0("ID=",.,"_",defaultlength,"nt_5UTR;Parent=",.,"_tx0") 
    }
    Strand <- pull(gff_join_row,Strand)
    if( Strand == "+" ) {
        Start.UTR <- pull(gff_join_row,Start.tx)
        if( is.na(Start.UTR) ) {
            Start.UTR <- pull(gff_join_row,Start.CDS) - defaultlength
        }
        End.UTR <- pull(gff_join_row,Start.CDS) - 1
    } else if( Strand == "-" ) {
        Start.UTR <- pull(gff_join_row,End.CDS) + 1
        End.UTR <- pull(gff_join_row,End.tx)
        if( is.na(End.UTR) ) {
            End.UTR <- pull(gff_join_row,End.CDS) + defaultlength
        }
    }
    tibble( Gene=pull(gff_join_row,Gene),
            Seqid=pull(gff_join_row,Seqid),
            Source=pull(gff_join_row,Source),
            Type="five_prime_UTR",
            Start=Start.UTR,
            End=End.UTR,
            Score=NA,
            Strand=Strand,
            Phase=NA,
            Attr=Attr.UTR)
}

assign_3UTR <- function(gff_join_row,defaultlength=125) {
    Attr.tx <- pull(gff_join_row,Attr.tx)
    if( !is.na(Attr.tx) ) {
        Attr.UTR <- Attr.tx %>%
            str_extract("ID=[\\w-_]+") %>%
            str_remove("ID=") %>% 
            paste0("ID=",.,"_3UTR;Parent=",.) 
    } else {
        Attr.UTR <- pull(gff_join_row,Gene) %>%
            paste0("ID=",.,"_",defaultlength,"nt_3UTR;Parent=",.,"_tx0")
    }
    Strand <- pull(gff_join_row,Strand)
    if( Strand == "-" ) {
        Start.UTR <- pull(gff_join_row,Start.tx)
        if( is.na(Start.UTR) ) {
            Start.UTR <- pull(gff_join_row,Start.CDS) - defaultlength
        }
        End.UTR <- pull(gff_join_row,Start.CDS) - 1
    } else if( Strand == "+" ) {
        Start.UTR <- pull(gff_join_row,End.CDS) + 1
        End.UTR <- pull(gff_join_row,End.tx)
        if( is.na(End.UTR) ) {
            End.UTR <- pull(gff_join_row,End.CDS) + defaultlength
        }
    }
    tibble( Gene=pull(gff_join_row,Gene),
            Seqid=pull(gff_join_row,Seqid),
            Source=pull(gff_join_row,Source),
            Type="three_prime_UTR",
            Start=Start.UTR,
            End=End.UTR,
            Score=NA,
            Strand=Strand,
            Phase=NA,
            Attr=Attr.UTR)
}

extend_feature <- function(gff_base,IDextl="ID=",IDextr="_ext",
                           ext5prime=25,ext3prime=125,extunstranded=0,
                           type="extension") {
    # extends a stranded feature in a gff_base by defined amounts
    # assumes there is a "Gene" column for ID, not ideal. 
    # This is very slow, maybe worth profiling?
    Strand <- pull(gff_base,Strand)
    if( Strand == "+" ) {
        Start <- pull(gff_base,Start) - ext5prime
        End   <- pull(gff_base,End)   + ext3prime
    } else if( Strand == "-" ) {
        Start <- pull(gff_base,Start) - ext3prime
        End   <- pull(gff_base,End)   + ext5prime
    } else {
        Start <- pull(gff_base,Start) - extunstranded
        End   <- pull(gff_base,End)   + extunstranded
    }
    tibble( Gene=pull(gff_base,Gene),
            Seqid=pull(gff_base,Seqid),
            Source=pull(gff_base,Source),
            Type=type,
            Start=Start,
            End=End,
            Score=NA,
            Strand=Strand,
            Phase=NA,
            Attr=paste0( IDextl, pull(gff_base,Gene), IDextr)
            )
}

union_gff_groups <- function(gff_in,...,type="Union") {
    gff_in %>%
        group_by(...) %>% 
        summarise(
            Seqid=unique(Seqid),
            Source=unique(Source),
            Type=type,
            Start=min(Start),
            End=max(End),
            Score=NA,
            Strand=unique(Strand),
            Phase=NA,
            Attr=unique(Attr)
        )
}
```

# Load data from gffs

```{r load_gffs}
gff_SGDall <- 
    here::here("data","saccharomyces_cerevisiae_R64-2-1_20150113.gff") %>%
    read_gff()

gff_CDS <- 
    gff_SGDall %>% 
    filter(Type=="CDS") %>%
    mutate(Gene = gene_from_attr(Attr)) 

gff_5pintrons <- 
    gff_SGDall %>% 
    filter(Type=="five_prime_UTR_intron") %>%
    mutate(Gene = gene_from_attr(Attr,after="_five_prime_UTR_intron")) 

gff_txs <- 
    here::here("data","longest_full-ORF_transcripts_ypd.gff3") %>%
    read_gff() %>%
    mutate(Gene = gene_from_attr(Attr,before="ID=")) 
```

## Which genes have 5'UTR introns? Which have longest annotations?

We need to address 5'UTR introns in particular becase several genes of interest (ribosomal proteins and Ssd1 targets) have 5'UTR introns, and we want to get their sequences right.

```{r check_5pintrons,dependson="load_gffs"}
gff_5pintrons %>%
    dplyr::mutate( longest_tx_annot = ( Gene %in% gff_txs$Gene) ) %>%
    dplyr::select(Gene,longest_tx_annot,Attr) %>%
    print(n=30)
```

```{r write_5pintrons,dependson="load_gffs"}
outfile_5UTR_introns <-
    here::here("out","five_prime_UTR_introns_R64-2-1_20150113.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=outfile_5UTR_introns)
write_gff(gff_5pintrons,
          path=outfile_5UTR_introns,
          append=TRUE)
```


Only one gene, YJL130C/URA2, has an annotated 5'UTR intron while not being in the longest_tx list. We can safely get the annotation wrong for this one gene.


# Join CDS annotation

```{r gff_join,dependson="load_gffs"}

# create the primary transcript region 
# only for nuclear verified ORFs
# this should differ only for CDS introns.
gff_verCDSunion <- 
    gff_CDS %>%
    filter( Seqid != "chrmt", str_detect(Attr,"Verified")) %>%
    union_gff_groups(Gene,type="primary_transcript_region")

gff_join_vCDS_txs <- 
    gff_verCDSunion %>%
    left_join(select(gff_txs, Gene, Seqid, Start, End, Strand,Attr),
              by=c("Gene","Seqid","Strand"), suffix=c(".CDS",".tx")) 
```

# Assign 5' UTRs, although keeping introns

There are 24 annotated 5'UTR introns in R64-2-1 which would ideally be removed from 5'UTRs. First, we keep them.

```{r gff_5UTRs_keepintrons,dependson="gff_join"}
gff_5UTRs_keepintrons <- 
    gff_join_vCDS_txs %>%
    group_by(Gene) %>%
    do(assign_5UTR(.))  %>%
    ungroup %>%
    mutate(Source="yeastutrgff")

outfile_5UTRs_keepintrons <- here::here("out","longest_five_prime_UTRs_keepintrons.gff")

write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=outfile_5UTRs_keepintrons)

write_gff(x=gff_5UTRs_keepintrons,
          path=outfile_5UTRs_keepintrons,
          append=TRUE)
```

# Assign 3' UTRs

```{r gff_3UTRs,dependson="gff_join"}
gff_3UTRs <- 
    gff_join_vCDS_txs %>%
    group_by(Gene) %>%
    do(assign_3UTR(.)) %>%
    ungroup %>%
    mutate(Source="yeastutrgff")

outfile_3UTRs <- here::here("out","longest_three_prime_UTRs.gff")

write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=outfile_3UTRs)

write_gff(x=gff_3UTRs,
          path=outfile_3UTRs,
          append=TRUE)
```

## Remove introns from 5'UTRs using bedtools

```{bash bedtools_subtract_5UTRintrons}
bedtools subtract -s \
-a ../out/longest_five_prime_UTRs_keepintrons.gff \
-b ../out/five_prime_UTR_introns_R64-2-1_20150113.gff > \
../out/longest_five_prime_UTRs_nointrons.gff
```

```{r load_gff_5UTRs_nointrons,dependson=c("gff_5UTRs_keepintrons","bedtools_subtract_5UTRintrons")}
gff_5UTRs_nointrons <- 
    here::here("out","longest_five_prime_UTRs_nointrons.gff") %>%
    read_gff()
```

# CDSs with default-length 25nt 5UTRs and 125nt 3'UTRs

```{r gff_CDS_extend,dependson="gff_join"}
gff_verCDSext <- 
    gff_verCDSunion %>%
    group_by(Gene) %>% 
    do(extend_feature(.,ext5prime=25,ext3prime=125,
                      IDextr="_tx0",type="primary_transcript")) %>%
    mutate(Source="yeastutrgff")

outfile_CDSext <- here::here(
    "out",
    "transcripts_fixedlength_25nt5p_125nt3p_UTRs.gff")

write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=outfile_CDSext)

write_gff(x=gff_verCDSext,
          path=outfile_CDSext,
          append=TRUE)
```

## Now make transcripts which are longest if we have data, fixed-length if we don't

```{r gff_longest_or_CDSext,dependson="gff_CDS_extend"}
genes_verified_no_longesttx <- 
    setdiff(gff_verCDSext$Gene,gff_txs$Gene)

gff_longest_or_CDSext <- 
    bind_rows(gff_txs,
              gff_verCDSext %>% 
                  filter(
                     Gene %in% genes_verified_no_longesttx)
              ) %>% 
    arrange(Gene)

outfile_longest_or_CDSext <- here::here(
    "out",
    "longest_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts.gff")

write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=outfile_longest_or_CDSext)

write_gff(x=gff_longest_or_CDSext,
          path=outfile_longest_or_CDSext,
          append=TRUE)
```

## Remove introns from CDS?

```{r gff_introns}
gff_allnucmRNAintrons <- 
    gff_SGDall %>% 
    filter(str_detect(Type,"intron"),str_detect(Attr,"Parent=Y")) %>%
    mutate(Gene = gene_from_attr(Attr,after="_intron")) 

outfile_nucmRNA_introns <-
    here::here("out","nuclear_mRNA_introns_R64-2-1_20150113.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today()),
        "##all introns from nuclear chromosome mRNAs, including 5'UTR introns"),
    path=outfile_nucmRNA_introns)
write_gff(gff_allnucmRNAintrons,
          path=outfile_nucmRNA_introns,
          append=TRUE)
```

## Remove introns from longest transcripts using bedtools

```{bash bedtools_subtract_allintrons}
bedtools subtract -s \
-a ../out/longest_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts.gff \
-b ../out/nuclear_mRNA_introns_R64-2-1_20150113.gff > \
../out/longest_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts_nointrons.gff
```

# Collected gff with CDS, introns, coding transcripts, UTRs, only

```{r gff_collected}
gff_collected <- 
    bind_rows( 
        gff_SGDall %>%
            filter( Type %in% 
                        c("chromosome",
                          "rRNA_gene","ncRNA_gene","telomerase_RNA_gene",
                          "snoRNA_gene","snRNA_gene","tRNA_gene")
            ),
        gff_allnucmRNAintrons,
        gff_5UTRs_nointrons,
        gff_3UTRs,
        gff_CDS
    ) %>%
    dplyr::arrange(Seqid,Start,desc(End))

outfile_gff_collected <-
    here::here("out","gff_Chromosomes_ncRNAs_mRNAparts.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today()),
        "##selected features: chromosomes, ncRNA, CDS, 5'UTRs, 3'UTRs, introns"),
    path=outfile_gff_collected)
write_gff(gff_collected,
          path=outfile_gff_collected,
          append=TRUE)

```

## gff with chromosomes only

```{r gff_chromosomesMito}
gff_chromosomesMito <- 
        gff_SGDall %>%
            filter( Type == "chromosome") %>%
    mutate(Seqid = str_replace(Seqid,"chrmt","chrMito"),
           Attr = str_replace_all(Attr,"chrmt","chrMito"))
    

outfile_gff_chromosomesMito <-
    here::here("out","gff_ChromosomesMito.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version SGD",
        paste("##date",lubridate::today()),
        "##chromosomes only with chrmt replaced by chrMito"),
    path=outfile_gff_chromosomesMito)
write_gff(gff_chromosomesMito,
          path=outfile_gff_chromosomesMito,
          append=TRUE)

```

# To Do

* Count Ssd1 reads on different kinds of features


