---
title: "Get sequences from Pelechano 2013 annotated transcripts etc."
author: "Edward Wallace, Edward.Wallace@ed.ac.uk"
date: "24 March 2020"
output: 
    html_document:
        toc: true
---

A script to get sequences of transcripts, UTRs, etc, mentioned in `annotateUTRs.Rmd`

# Try with bedtools; works on seqs but not on names

```{bash bedtools_txfasta, eval=FALSE}
bedtools getfasta -s -fullHeader \
-fi ../data/saccharomyces_cerevisiae_R64-2-1_20150113_chrnames.fsa \
-bed ../data/longest_full-ORF_transcripts_ypd.gff3 \
-fo ../out/longest_full-ORF_transcripts_keepintrons_ypd_getfasta.fasta
```

# Try with bioconductor

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(Biostrings)
library(rtracklayer)
# library(Rsamtools)
library(tidyverse)
library(here)
```

## Load genome fasta

```{r load_genome}

Scer_fastafile <- 
    here::here("data","saccharomyces_cerevisiae_R64-2-1_20150113_chrnames.fsa")

Scer_FaFile <- Rsamtools::FaFile(Scer_fastafile)
Scer_fasta <- 
    Biostrings::readDNAStringSet(Scer_fastafile)
```

## Get transcript sequences

```{r longest_tx_seqs,dependson="load_genome"}
granges_txs <- 
    here::here("data","longest_full-ORF_transcripts_ypd.gff3") %>%
    rtracklayer::readGFFAsGRanges() 

seqs_txs <- Biostrings::getSeq(Scer_FaFile,granges_txs) %>%
    magrittr::set_names(granges_txs$ID)

Biostrings::writeXStringSet(
    seqs_txs, 
    here::here("out",
               "longest_full-ORF_transcripts_keepintrons_ypd.fasta")
)
```

```{r longest_etc_tx_seqs,dependson="load_genome"}
granges_splicedtxs <- 
    here::here("out","longest_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts_nointrons.gff") %>%
    rtracklayer::readGFFAsGRanges() 

seqs_txexons <- Biostrings::getSeq(Scer_FaFile,granges_splicedtxs) # %>%
    # set_names(paste(granges_splicedtxs$ID,
    #                 chrom(granges_splicedtxs),
    #                 start(granges_splicedtxs),
    #                 sep="_")
    # )

concat_seqdf_startorder <- function(seqdf) {
    ID = unique(seqdf$ID)
    Strand = unique(seqdf$strand)
    if( Strand == "+") {
       seqdfarr <- 
           dplyr::arrange(seqdf,start) 
    } else if ( Strand == "-" ) {
        seqdfarr <- 
            dplyr::arrange(seqdf,desc(start))
    }
    tibble::tibble(Seq=paste0(seqdfarr$Seq,collapse = ""))
}

df_splicedtxs <- granges_splicedtxs %>%
    as.data.frame() %>% 
    tibble::as_tibble() %>%
    dplyr::mutate(Seq = as.character(seqs_txexons))

# debug(concat_seqdf_startorder)
# df_splicedtx %>% dplyr::filter(ID=="YAL001C_tx0") %>% concat_seqdf_startorder
# df_splicedtx %>% dplyr::filter(ID=="YAL038W_id004") %>% concat_seqdf_startorder
# df_splicedtx %>% dplyr::filter(ID=="YIL123W_id002") %>% concat_seqdf_startorder

df_splicedtxs_seq <- 
    df_splicedtxs %>%
    dplyr::group_by(ID) %>%
    dplyr::do(concat_seqdf_startorder(.))

seqs_splicedtxs <- 
    df_splicedtxs_seq$Seq %>% 
    DNAStringSet() %>%
    magrittr::set_names(df_splicedtxs_seq$ID)

Biostrings::writeXStringSet(
    seqs_splicedtxs, 
    here::here("out",
               "longest_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts_nointrons.fasta")
)
```

```{r abundant_etc_tx_seqs,dependson="load_genome"}
granges_splicedtxs <- 
    here::here("out","abundant_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts_nointrons.gff") %>%
    rtracklayer::readGFFAsGRanges() 

seqs_txexons <- Biostrings::getSeq(Scer_FaFile,granges_splicedtxs) # %>%
    # set_names(paste(granges_splicedtxs$ID,
    #                 chrom(granges_splicedtxs),
    #                 start(granges_splicedtxs),
    #                 sep="_")
    # )

concat_seqdf_startorder <- function(seqdf) {
    ID = unique(seqdf$ID)
    Strand = unique(seqdf$strand)
    if( Strand == "+") {
       seqdfarr <- 
           dplyr::arrange(seqdf,start) 
    } else if ( Strand == "-" ) {
        seqdfarr <- 
            dplyr::arrange(seqdf,desc(start))
    }
    tibble::tibble(Seq=paste0(seqdfarr$Seq,collapse = ""))
}

df_splicedtxs <- granges_splicedtxs %>%
    as.data.frame() %>% 
    tibble::as_tibble() %>%
    dplyr::mutate(Seq = as.character(seqs_txexons))

# debug(concat_seqdf_startorder)
# df_splicedtx %>% dplyr::filter(ID=="YAL001C_tx0") %>% concat_seqdf_startorder
# df_splicedtx %>% dplyr::filter(ID=="YAL038W_id004") %>% concat_seqdf_startorder
# df_splicedtx %>% dplyr::filter(ID=="YIL123W_id002") %>% concat_seqdf_startorder

df_splicedtxs_seq <- 
    df_splicedtxs %>%
    dplyr::group_by(ID) %>%
    dplyr::do(concat_seqdf_startorder(.))

seqs_splicedtxs <- 
    df_splicedtxs_seq$Seq %>% 
    DNAStringSet() %>%
    magrittr::set_names(df_splicedtxs_seq$ID)

Biostrings::writeXStringSet(
    seqs_splicedtxs, 
    here::here("out",
               "abundant_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts_nointrons.fasta")
)
```

## Get 5'UTR sequences

```{r longest_5UTRs_keepintrons,dependson="load_genome"}
granges_5UTRs_keep <- 
    here::here("out","longest_five_prime_UTRs_keepintrons.gff") %>%
    rtracklayer::readGFFAsGRanges() 

seqs_5UTRs_keep <- Biostrings::getSeq(Scer_FaFile,granges_5UTRs_keep) %>%
    magrittr::set_names(granges_5UTRs_keep$ID)

Biostrings::writeXStringSet(
    seqs_5UTRs_keep, 
    here::here("out",
               "longest_five_prime_UTRs_keepintrons.fasta")
)
```

# To do:

* Make longest or alternately fixed-length-UTR transcript seqs.
* Make transcripts with introns removed?
* Make a version with just spliced 5'UTRs?

