---
title: "Subset transcript sequences for Ssd1 targets"
author: "Edward Wallace, Edward.Wallace@ed.ac.uk"
date: "24 March 2020"
output: 
    html_document:
        toc: true
---

A script to get transcripts targeted by Ssd1 from published data, then separate their transcript targets and do motif analysis on them.

Now updated to reflect verified ORFs only, longest transcripts when taken from Pelechano 2013, imputed fixed-length UTRs otherwise, and spliced for mature transcripts only.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(Biostrings)
library(here)
library(VennDiagram)
venn <- function(...) {
    venn.diagram(x=list(...),filename=NULL,margin=0.1) %>% grid.draw
}
```

```{r load_txs}
txs_fasta <- 
    Biostrings::readDNAStringSet(
        here::here("out",
                   "longest_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts_nointrons.fasta")
    )
names(txs_fasta) <- 
    names(txs_fasta) %>% 
    str_remove("_.*")

five_UTRs_fasta <- 
    Biostrings::readDNAStringSet(
        here::here("out",
                   "longest_five_prime_UTRs_keepintrons.fasta")
    )
names(five_UTRs_fasta) <- 
    names(five_UTRs_fasta) %>% 
    str_remove("_.*") 
```


# Lists of mRNAs enriched in Ssd1 binding

## Hogan 2008

Reading data from Hogan 2008 `https://doi.org/10.1371/journal.pbio.0060255.sd003`.

```{r Hogan_targets}
# hoganFDRsall <- 
#     here::here("data","Ssd1_targets","Hogan2008_FDR_Dataset_S2.txt") %>%
#     read_tsv(comment="#")
# 
# hoganFDRsall

ssd1_data_hogan2008 <-
    here::here("data","Ssd1_targets",
               "journal.pbio.0060255.sd003_Hogan2008_Dataset_S3.xls") %>%
    readxl::read_excel("Ssd1_sam",skip=1,
                       col_names = c("Gene",
                                     "Score",
                                     "Numerator",
                                     "Denominator",
                                     "FoldChange",
                                     "qvalue",
                                     "localfdr")
    )

ssd1_data_hogan2008

ssd1_targets_hogan2008 <- 
    ssd1_data_hogan2008 %>%
    filter(str_detect(Gene,"_ORF"),localfdr < 1) %>%
    pull(Gene) %>%
    str_remove("_.*") %>%
    unique
```

## Hogan 2008 target seqs

```{r Hogan_seqs,dependson="Hogan_targets"}
ssd1_targets_hogan2008_woseqs <- setdiff(ssd1_targets_hogan2008,names(txs_fasta))

ssd1_targets_hogan2008_seqs <-
    txs_fasta[intersect(ssd1_targets_hogan2008,names(txs_fasta))]

writeXStringSet(ssd1_targets_hogan2008_seqs,
                here::here(
                    "out","Ssd1_targets",
                    "ssd1_targets_hogan2008_localfdr5_longestetcspliced_txseqs.fasta"))
```

The following Ssd1 targets are not in the list of transcript sequences:

```{r dependson="Hogan_seqs"}
ssd1_targets_hogan2008_woseqs
```

That suggests that we should make educated guesses for those UTRs.

```{r Hogan_5UTR_seqs,dependson="Hogan_targets"}
ssd1_targets_hogan2008_5UTR_seqs <-
    five_UTRs_fasta[intersect(ssd1_targets_hogan2008,names(five_UTRs_fasta))]

writeXStringSet(ssd1_targets_hogan2008_5UTR_seqs,
                here::here(
                    "out","Ssd1_targets",
                    "ssd1_targets_hogan2008_localfdr5_5UTRs_keepintrons.fasta"))

writeXStringSet(ssd1_targets_hogan2008_5UTR_seqs[
  width(ssd1_targets_hogan2008_5UTR_seqs) > 15 ],
                here::here(
                    "out","Ssd1_targets",
                    "ssd1_targets_hogan2008_localfdr5_5UTRs_keepintrons_minwidth15.fasta"))
```

## Jansen 2009

Reading data from Jansen 2009.

Surprisingly, the gene systematic names in this table are not completely systematic.

```{r Jansen_targets}

ssd1_data_jansen2009 <-
    here::here("data","Ssd1_targets",
               "NIHMS158901-supplement-02.xls") %>%
    readxl::read_excel(sheet=1)

ssd1_data_jansen2009 

# Here we have to take some effort to get systematic names for all targets
ssd1_genic_targets_jansen2009 <- 
    ssd1_data_jansen2009 %>%
    filter(`inter/genic?`=="genic") %>%
    pull(sys_id)

ssd1_intergenic_targets_jansen2009 <- 
    c(ssd1_data_jansen2009 %>%
          filter(`inter/genic?`!="genic",`sys_id`==`left gene name`) %>%
          pull(`left_gene sys id`),
      ssd1_data_jansen2009 %>%
          filter(`inter/genic?`!="genic",`sys_id`==`right gene name`) %>%
          pull(`right_gene sys id`)
    )

ssd1_targets_jansen2009 <- unique( c(ssd1_genic_targets_jansen2009,
                                     ssd1_intergenic_targets_jansen2009) )
```

## Jansen 2009 target seqs

```{r Jansen_seqs,dependson="Jansen_targets"}
ssd1_targets_jansen2009_woseqs <- setdiff(ssd1_targets_jansen2009,names(txs_fasta))

ssd1_targets_jansen2009_seqs <-
    txs_fasta[intersect(ssd1_targets_jansen2009,names(txs_fasta))]

writeXStringSet(ssd1_targets_jansen2009_seqs,
                here::here(
                    "out","Ssd1_targets",
                    "ssd1_targets_jansen2009_localfdr5_longestetcspliced_txseqs.fasta"))
```

The following Ssd1 targets are not in the list of transcript sequences:

```{r dependson="Jansen_seqs"}
ssd1_targets_jansen2009_woseqs
```

These are dubious or otherwise non-verified ORFs.

## Hose 2020

```{r Hose_targets}
ssd1_data_hose2020 <-
    here::here("data","Ssd1_targets",
               "elife-52063-supp2-v2_Hose2020.xlsx") %>%
    readxl::read_excel(range = "A2:E6715",
                       col_names = c("Gene",
                                     "Annotation",
                                     "Bound",
                                     "RNA_DE",
                                     "Prot_DE")
    )

ssd1_data_hose2020

ssd1_targets_hose2020 <- 
    ssd1_data_hose2020 %>%
    filter(Bound=="yes") %>%
    pull(Gene)
```

## Hose 2020_seqs

```{r Hose_seqs,dependson="Hose_targets"}
ssd1_targets_hose2020_woseqs <- setdiff(ssd1_targets_hose2020,names(txs_fasta))

ssd1_targets_hose2020_seqs <-
    txs_fasta[intersect(ssd1_targets_hose2020,names(txs_fasta))]

writeXStringSet(ssd1_targets_hose2020_seqs,
                here::here(
                    "out","Ssd1_targets",
                    "ssd1_targets_hose2020_fdr5_longestetcspliced_txseqs.fasta"))
```

The following Ssd1 targets are not in the list of transcript sequences:

```{r dependson="Hose_seqs"}
ssd1_targets_hose2020_woseqs
```

```{r Hose_5UTR_seqs,dependson="Hogan_targets"}
ssd1_targets_hose2020_5UTR_seqs <-
    five_UTRs_fasta[intersect(ssd1_targets_hose2020,names(five_UTRs_fasta))]

writeXStringSet(ssd1_targets_hose2020_5UTR_seqs,
                here::here(
                    "out","Ssd1_targets",
                    "ssd1_targets_hose2020_fdr5_5UTRs_keepintrons.fasta"))

writeXStringSet(ssd1_targets_hose2020_5UTR_seqs[
  width(ssd1_targets_hose2020_5UTR_seqs) > 15 ],
                here::here(
                    "out","Ssd1_targets",
                    "ssd1_targets_hose2020_fdr5_5UTRs_keepintrons_minwidth15.fasta"))
```

# Venn diagram of intersections

Ugly yet effective.

```{r venn_hogan_jansen_hose,dependson=c("Hogan_targets","Jansen_targets","Hose_targets"),fig.width=4,fig.height=3.8}
venn(Hogan=ssd1_targets_hogan2008,
     Jansen=ssd1_targets_jansen2009,
     Hose=ssd1_targets_hose2020)
```


## Bayne targets

Removing dubious ORFs ets that don't have assigned gene names.

```{r bayne_targets}
ssd1_data_bayne2020 <-
    here::here("data","Ssd1_targets",
               "Ssd1_targets_TPMratio5x_30C_EW1Apr2020.txt") %>%
    readr::read_tsv()

ssd1_data_bayne2020_42C <-
    here::here("data","Ssd1_targets",
               "Ssd1_targets_TPMratio5x_42C_EW1Apr2020.txt") %>%
    readr::read_tsv()


ssd1_targets_bayne2020 <- 
    ssd1_data_bayne2020 %>%
    filter(!is.na(Name)) %>%
    pull(Gene)

ssd1_targets_bayne2020_42C <- 
    ssd1_data_bayne2020_42C %>%
    filter(!is.na(Name)) %>%
    pull(Gene)
```


# Venn diagram of intersections with all datasets

Ugly yet effective.

```{r venn_hogan_jansen_hose_bayne,dependson=c("Hogan_targets","Jansen_targets","Hose_targets"),fig.width=5,fig.height=3.8}
venn(Hogan2008=ssd1_targets_hogan2008,
     Jansen2009=ssd1_targets_jansen2009,
     Hose2020=ssd1_targets_hose2020,
     Bayne5x30C=ssd1_targets_bayne2020,
     Bayne5x42C=ssd1_targets_bayne2020_42C)
```

### What are the genes common to all datasets?

First make a gene-name data frame to show nice gene names:

```{r load_genenames,warning=FALSE,message=FALSE}
gene_name_df <- 
  here::here("data","saccharomyces_cerevisiae_R64-2-1_20150113.gff") %>%
    read_tsv(col_names=FALSE,comment="#") %>%
    filter(X3=="gene") %>%
    transmute(Gene = str_extract(X9,"Name=[\\w-]+") %>%
                str_remove("Name="),
              Name = str_extract(X9,"gene=[\\w]+") %>%
                str_remove("gene="))
```

Then find the intersection. Generously, we are taking 5x enriched targets in Bayne 2020 at either 30 or 42C.

```{r Ssd1_10commontargets,results="show"}
ssd1_targets_10common <- 
    union(ssd1_targets_bayne2020, ssd1_targets_bayne2020_42C) %>%
    intersect(ssd1_targets_hogan2008) %>%
    intersect(ssd1_targets_jansen2009) %>%
    intersect(ssd1_targets_hose2020)
    
filter(gene_name_df,Gene %in% ssd1_targets_10common)
```


## Smaller Venn diagrams of intersections with our data

Only including the more conservative Hogan 2008 and Jansen 2009.

### Hogan 2008 vs Jansen 2009 vs Bayne 30C

```{r venn_hogan_jansen_bayne,dependson=c("Hogan_targets","Jansen_targets","Hose_targets"),fig.width=4,fig.height=3.8}
venn(Hogan2008=ssd1_targets_hogan2008,
     Jansen2009=ssd1_targets_jansen2009,
     # Hose2020=ssd1_targets_hose2020,
     Bayne5x30C=ssd1_targets_bayne2020)
```

### Hogan vs 30C vs 42C

```{r venn_hogan_bayne_2temps,dependson=c("Hogan_targets","Jansen_targets","Hose_targets"),fig.width=4,fig.height=3.8}
venn(Hogan2008=ssd1_targets_hogan2008,
     #Jansen2009=ssd1_targets_jansen2009,
     # Hose2020=ssd1_targets_hose2020,
     Bayne5x30C=ssd1_targets_bayne2020,
     Bayne5x42C=ssd1_targets_bayne2020_42C)
```

# To do:

* MEME analysis in Differential Enrichment mode, Ssd1 targets vs other